src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs

```diff
@@ -0,0 +1,243 @@
+                        {
+                            kvp.Value.Last = DateTime.MinValue;
+                        }
+
+                        while (this.DeliverNextMessage(kvp.Value, out var next))
+                        {
+                            yield return next;
+                        }
+
+                        if (kvp.Value.Last == DateTime.MinValue
+                            && (kvp.Value.Buffered == null || kvp.Value.Buffered.Count == 0))
+                        {
+                            (emptyReceiveBuffers ?? (emptyReceiveBuffers = new List<string>())).Add(kvp.Key);
+                        }
+                    }
+
+                    if (emptyReceiveBuffers != null)
+                    {
+                        foreach (var t in emptyReceiveBuffers)
+                        {
+                            this.Received.Remove(t);
+                        }
+                    }
+
+                    if (this.Received.Count == 0)
+                    {
+                        this.Received = null;
+                    }
+                }
+            }
+
+            // Messages older than the reorder window are not participating.
+            if (message.Timestamp < this.ReceiveHorizon)
+            {
+                // Just pass the message through.
+                yield return message;
+                yield break;
+            }
+
+            ReceiveBuffer receiveBuffer;
+
+            if (this.Received == null)
+            {
+                this.Received = new Dictionary<string, ReceiveBuffer>()
+                {
+                    { message.ParentInstanceId,  receiveBuffer = new ReceiveBuffer() },
+                };
+            }
+            else if (!this.Received.TryGetValue(message.ParentInstanceId, out receiveBuffer))
+            {
+                this.Received[message.ParentInstanceId] = receiveBuffer = new ReceiveBuffer();
+            }
+
+            if (message.Timestamp <= receiveBuffer.Last)
+            {
+                // This message was already delivered, it's a duplicate
+                yield break;
+            }
+
+            if (message.Predecessor > receiveBuffer.Last
+                && message.Predecessor >= this.ReceiveHorizon)
+            {
+                // this message is waiting for a non-delivered predecessor in the window, buffer it
+                if (receiveBuffer.Buffered == null)
+                {
+                    receiveBuffer.Buffered = new SortedDictionary<DateTime, RequestMessage>();
+                }
+
+                receiveBuffer.Buffered[message.Timestamp] = message;
+            }
+            else
+            {
+                yield return message;
+
+                receiveBuffer.Last = message.Timestamp >= this.ReceiveHorizon ? message.Timestamp : DateTime.MinValue;
+
+                while (this.DeliverNextMessage(receiveBuffer, out var next))
+                {
+                    yield return next;
+                }
```

Comment:
Should we update the ReceiveHorizon here?

Reply:
Not necessary. The message we are delivering was already inspected earlier (before we inserted it into the reorder buffer) so the ReceiveHorizon was already updated back then.